File: frontend/src/types/user-model.ts
/**
 * User Model Types - 用户模型类型
 */

export type ModelType = 'text' | 'image' | 'image_gen' | 'video'

export interface UserModelConfig {
  context_window?: number
  supports_vision?: boolean
  supports_tools?: boolean
  supports_reasoning?: boolean
  max_tokens?: number
  input_price?: number
  output_price?: number
  description?: string
}

/** 模型连通性测试三态：未测过 → null。 */
export type ModelTestStatus = 'success' | 'failed' | null

export interface UserModel {
  id: string
  user_id: string | null
  anonymous_user_id: string | null
  display_name: string
  provider: string
  model_id: string
  api_key_masked: string | null
  has_api_key: boolean
  api_base: string | null
  model_types: ModelType[]
  config: UserModelConfig | null
  is_active: boolean
  is_system: boolean
  /** 上次连通性测试结果，未测过为 null */
  last_test_status: ModelTestStatus
  /** 上次连通性测试时间（ISO 8601），未测过为 null */
  last_tested_at: string | null
  /** 上次失败/不支持时的说明；成功或未测过为 null */
  last_test_reason: string | null
  created_at: string | null
  updated_at: string | null
}

export interface SystemModel {
  id: string
  display_name: string
  provider: string
  model_id: string
  model_types: ModelType[]
  is_system: true
  config: UserModelConfig | null
}

/** 默认模型信息（用于展示「默认（模型名）」） */
export interface DefaultModelInfo {
  id: string
  display_name: string
}

export interface AvailableModelsResponse {
  system_models: SystemModel[]
  user_models: UserModel[]
  /** 文本能力默认模型（未选时展示） */
  default_for_text?: DefaultModelInfo
  /** 视觉能力默认模型（未选时展示） */
  default_for_vision?: DefaultModelInfo
  /** 图像生成默认模型（未选时展示） */
  default_for_image_gen?: DefaultModelInfo
}

export interface CreateUserModelBody {
  display_name: string
  provider: string
  model_id: string
  api_key?: string | null
  api_base?: string | null
  model_types?: ModelType[]
  config?: UserModelConfig | null
}

export interface UpdateUserModelBody {
  display_name?: string
  provider?: string
  model_id?: string
  api_key?: string | null
  api_base?: string | null
  model_types?: ModelType[]
  config?: UserModelConfig | null
  is_active?: boolean
}

export interface TestConnectionResult {
  success: boolean
  message: string
  model: string
  /** 与 ORM 落库字段同名，便于乐观更新 */
  status?: ModelTestStatus
  /** ISO 8601；与 last_tested_at 同语义 */
  tested_at?: string
  /** 与 ORM last_test_reason 一致；成功为 undefined/null */
  reason?: string | null
  response_preview?: string
}

export const MODEL_PROVIDERS = [
  { id: 'openai', name: 'OpenAI' },
  { id: 'anthropic', name: 'Anthropic' },
  { id: 'deepseek', name: 'DeepSeek' },
  { id: 'dashscope', name: '阿里云 (DashScope)' },
  { id: 'zhipuai', name: '智谱 AI' },
  { id: 'volcengine', name: '火山引擎' },
  { id: 'custom', name: '自定义' },
] as const

export const MODEL_TYPE_LABELS: Record<ModelType, string> = {
  text: '文本',
  image: '图片理解',
  image_gen: '图片生成',
  video: '视频',
}

File: backend/domains/agent/infrastructure/models/user_model.py
"""
User Model - 用户自定义模型配置

用户可配置自带 API Key 的 LLM 模型，其他用户不可见。
模型类型：text（文本）、image（图片）、video（视频），支持多选。
"""

from datetime import datetime
from typing import Any
import uuid

from sqlalchemy import Boolean, DateTime, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import ARRAY, JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from libs.orm.base import BaseModel, OwnedMixin


class UserModel(BaseModel, OwnedMixin):
    """用户自定义模型"""

    __tablename__ = "user_models"

    user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    anonymous_user_id: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
        index=True,
    )

    display_name: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        comment="显示名称",
    )
    provider: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        comment="提供商: openai/deepseek/dashscope/anthropic/zhipuai/volcengine/custom",
    )
    model_id: Mapped[str] = mapped_column(
        String(200),
        nullable=False,
        comment="模型标识, 如 gpt-4o, deepseek-chat",
    )
    api_key_encrypted: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
        comment="加密存储的 API Key (Fernet)",
    )
    api_base: Mapped[str | None] = mapped_column(
        String(500),
        nullable=True,
        comment="自定义 API 端点",
    )
    model_types: Mapped[list[str]] = mapped_column(
        ARRAY(String(20)),
        nullable=False,
        comment="模型类型: text, image, video",
    )
    config: Mapped[dict[str, Any] | None] = mapped_column(
        JSONB,
        nullable=True,
        comment="扩展配置 {context_window, supports_vision, supports_tools, max_tokens, ...}",
    )
    is_active: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False,
        comment="是否启用",
    )
    last_test_status: Mapped[str | None] = mapped_column(
        String(20),
        nullable=True,
        comment="上次连通性测试结果: success / failed / NULL=未测过",
    )
    last_tested_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        comment="上次连通性测试时间",
    )
    last_test_reason: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
        comment="上次连通性测试说明（失败原因等）；成功时为 NULL",
    )

    def __repr__(self) -> str:
        return f"<UserModel {self.id} {self.display_name} ({self.provider}/{self.model_id})>"
